codegen/codegen.go

```diff
@@ -315,35 +319,35 @@ func (cg *CodeGen) EmitStringChainStmt(ctx context.Context, scope *parser.Scope,
 	}
 }
 
-func (cg *CodeGen) EmitFilesystemBlock(ctx context.Context, scope *parser.Scope, stmts []*parser.Stmt, ac aliasCallback) (llb.State, error) {
-	v, err := cg.EmitBlock(ctx, scope, parser.Filesystem, stmts, ac)
+func (cg *CodeGen) EmitFilesystemBlock(ctx context.Context, scope *parser.Scope, stmts []*parser.Stmt, ac aliasCallback, chainStart interface{}) (llb.State, error) {
+	v, err := cg.EmitBlock(ctx, scope, parser.Filesystem, stmts, ac, chainStart)
 	return v.(llb.State), err
 }
 
-func (cg *CodeGen) EmitStringBlock(ctx context.Context, scope *parser.Scope, stmts []*parser.Stmt) (string, error) {
-	v, err := cg.EmitBlock(ctx, scope, parser.Str, stmts, noopAliasCallback)
+func (cg *CodeGen) EmitStringBlock(ctx context.Context, scope *parser.Scope, stmts []*parser.Stmt, chainStart interface{}) (string, error) {
+	v, err := cg.EmitBlock(ctx, scope, parser.Str, stmts, noopAliasCallback, chainStart)
 	if v == nil {
 		return "", err
 	}
 	return v.(string), err
 }
 
-func (cg *CodeGen) EmitFuncLit(ctx context.Context, scope *parser.Scope, lit *parser.FuncLit, op string, ac aliasCallback) (interface{}, error) {
+func (cg *CodeGen) EmitFuncLit(ctx context.Context, scope *parser.Scope, lit *parser.FuncLit, op string, ac aliasCallback, chainStart interface{}) (interface{}, error) {
 	switch lit.Type.Primary() {
 	case parser.Int, parser.Bool:
 		panic("unimplemented")
 	case parser.Filesystem:
-		return cg.EmitFilesystemBlock(ctx, scope, lit.Body.NonEmptyStmts(), ac)
+		return cg.EmitFilesystemBlock(ctx, scope, lit.Body.NonEmptyStmts(), ac, chainStart)
 	case parser.Str:
-		return cg.EmitStringBlock(ctx, scope, lit.Body.NonEmptyStmts())
+		return cg.EmitStringBlock(ctx, scope, lit.Body.NonEmptyStmts(), chainStart)
 	case parser.Option:
-		return cg.EmitOptions(ctx, scope, op, lit.Body.NonEmptyStmts(), ac)
+		return cg.EmitOptions(ctx, scope, op, lit.Body.NonEmptyStmts(), ac, chainStart)
 	default:
 		return nil, errors.WithStack(ErrCodeGen{lit, errors.Errorf("unknown func lit")})
 	}
 }
 
-func (cg *CodeGen) EmitWithOption(ctx context.Context, scope *parser.Scope, parent *parser.CallStmt, with *parser.WithOpt, ac aliasCallback) (opts []interface{}, err error) {
+func (cg *CodeGen) EmitWithOption(ctx context.Context, scope *parser.Scope, parent *parser.CallStmt, with *parser.WithOpt, ac aliasCallback, chainStart interface{}) (opts []interface{}, err error) {
```

Comment:
`EmitWithOption` shouldn't not have `chainStart`, it's a new context.

Reply:
fixed